bpf: fix error handling for invoke_tailcall_if()

julianwiedmann opened the pull request:
`invoke_tailcall_if()` either inlines the target function, or calls it via
tail-call. In the inline variant, the `return` causes us to skip over any
subsequent error handling (eg drop notification) in the caller.

Rework the macro so that it properly assigns the `ret` value.

Also fix up one user that didn't handle returned errors.

julianwiedmann committed:
bpf: fix error handling for invoke_tailcall_if()

invoke_tailcall_if() either inlines the target function, or calls it via
tail-call. In the inline variant, the `return` causes us to skip over any
subsequent error handling (eg drop notification) in the caller.

Rework the macro so that it properly assigns the `ret` value.

Suggested-by: Anton Protopopov <aspsk@isovalent.com>
Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

aspsk commented on bpf/bpf_host.c:
```diff
@@ -409,11 +409,13 @@ tail_handle_ipv6(struct __ctx_buff *ctx, const bool from_host)
 			invoke_tailcall_if(is_defined(ENABLE_HOST_FIREWALL),
 					   CILIUM_CALL_IPV6_CONT_FROM_HOST,
-					   tail_handle_ipv6_cont_from_host);
+					   tail_handle_ipv6_cont_from_host,
```
How about
```
ret = invoke_tailcall_if(is_defined(ENABLE_HOST_FIREWALL),
					   CILIUM_CALL_IPV6_CONT_FROM_HOST,
					   tail_handle_ipv6_cont_from_host);
```
?

julianwiedmann replied:
That would be easier to read, but I couldn't think of a pretty way to make it work for when the macro expands into a tail-call.

aspsk replied:
But then it just doesn't return, so something like
```
#define __invoke_tailcall_if_1(NAME, FUNC)	({			\
		ep_tail_call(ctx, NAME);				\
                DROP_MISSED_TAIL_CALL;				\
})
```
should work

julianwiedmann replied:
yes that should do the trick, thanks :pray:

julianwiedmann committed:
bpf: lxc: add error handling for invoke_tailcall_if() in CT tail-call

If the invoked function is inlined, it can return error that we need to
handle. If the function is tail-called, we still need to handle the
DROP_MISSED_TAIL_CALL.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>
